### DISCUSSION TITLE
Kubernetes Operator

### DISCUSSION THREAD

**Discussion by sbordeyne** [kanidm/kanidm#2302] [General]
Hello all, I'm creating a new discussion here, since it seems more appropriate to centralize information (rather than on issues / pull requests)

I'm working on a kubernetes operator to make kanidm "cloud native" (or so the buzzword say). The goals are to:

- Automate deployment of kanidm on a kubernetes cluster with either a single deployment or HA configuration (eventually with autoscaling if I'm determined enough)
- Automate creation of service accounts, accounts, groups, subgroups, oauth clients, radius, ldap config and password badlisting through kubernetes custom resources

The repo is here : https://github.com/sbordeyne/kanidm-operator 

If the operator gets stable enough it should probably be migrated to the kanidm github organisation (which would make it official-ish), but it's not yet ready for that.

The operator is written in Python (sorry gophers), since I find it more convenient (and the `kubernetes` and `kopf` packages are good enough for the use case).

Don't hesitate to ask any questions. Feature requests related to the operator should be done in issues on the operator's repository. As a side note, and for convenience, the first version of the operator will depend on `cert-manager` to be available in the cluster the kanidm operator will be deployed on. It simplifies certificate handling greatly, and is well supported.

If anyone wishes to help, it's appreciated, @yaleman is already working on implementing missing features of the `kanidm` python client. Help in the form of testing environments on different cloud platforms is also great, I for one know that GCP has special GKE specific kubernetes resource definitions such as `BackendConfig` and `FrontendConfig` which may be required (especially for HA when the time comes).

**Comment by Firstyear** [comment]
For now, you don't want to do autoscaling because there are some limitations in replication (especially node removal) that could cause issues. So avoid that for now.
